### ISSUE TITLE
Floating hover/preview windows get stuck open under neovim if Preview window is open

### ISSUE THREAD

**Issue by cristoper** [prabirshrestha/vim-lsp#819]
I'm using neovim 0.4.0 and vim-lsp `a943fe5`. If the Preview window is open (which is likely during completion if completeopt+=preview), then when continuing to edit after LspHover, the Preview window is closed instead of the floating window showing the hover info. The floating window is then stuck in the middle of the screen and must be focused and closed manually.

I do not have this problem in vim 8.2.

I can fix it for me by using `nvim_win_close` instead of `pclose` in lsp#ui#vim#output#closepreview:

```diff
diff --git a/autoload/lsp/ui/vim/output.vim b/autoload/lsp/ui/vim/output.vim
index be4b575..7d4c8f7 100644
--- a/autoload/lsp/ui/vim/output.vim
+++ b/autoload/lsp/ui/vim/output.vim
@@ -15,12 +15,11 @@ function! lsp#ui#vim#output#closepreview() abort
         " Don't close if window got focus
         return
     endif
-    "closing floats in vim8.1 must use popup_close() (nvim could use nvim_win_close but pclose
-    "works)
+    "closing floats in vim8.1 must use popup_close()
     if s:use_vim_popup && s:winid
         call popup_close(s:winid)
     else
-        pclose
+        call nvim_win_close(s:winid, 0)
     endif
     let s:winid = v:false
     let s:preview_data = v:false
```

**Comment by cristoper** [comment]
I was getting errors about invalid winid with the above patch, so I'm using this now (just suppressing the errors with `silent!`). but I don't really know what I'm doing:

```diff
diff --git a/autoload/lsp/ui/vim/output.vim b/autoload/lsp/ui/vim/output.vim
index be4b575..fdd61a7 100644
--- a/autoload/lsp/ui/vim/output.vim
+++ b/autoload/lsp/ui/vim/output.vim
@@ -15,10 +15,11 @@ function! lsp#ui#vim#output#closepreview() abort
         " Don't close if window got focus
         return
     endif
-    "closing floats in vim8.1 must use popup_close() (nvim could use nvim_win_close but pclose
-    "works)
+    "closing floats in vim8.1 must use popup_close()
     if s:use_vim_popup && s:winid
         call popup_close(s:winid)
+    elseif s:winid
+            silent! call nvim_win_close(s:winid, 0)
     else
         pclose
     endif
```

**Comment by cristoper** [comment]
I've reproduced this issue with a minimal .vimrc (below) and on two computers (nvim 0.4.0 on macos and nvim 0.4.3 on linux).

In the first screenshot below I am in a Python file (using pyls) and I have typed `aifc.<C-x><C-o>` selected a method to complete using vim-lsp's omnifunc, and then typed an opening paren `(`. The preview window is open (because `completeopt=menuone,preview`) and vim-lsp is showing the Hover info in a floating window.

By the way, is there any way to disable vim-lsp's automatic hover in insert mode?

![Screenshot_2020-05-23_11-23-34](https://user-images.githubusercontent.com/331548/82736634-88d18680-9ce8-11ea-9cde-be1ea8a466e0.png)

In the next screenshot I have typed `)` and the preview window is closed but the floating window then becomes stuck in the middle of the main window. I have to type `C-w C-c` to close it manually:

![Screenshot_2020-05-23_11-24-37](https://user-images.githubusercontent.com/331548/82736724-0eedcd00-9ce9-11ea-9ee1-dfacb53c3d7e.png)

The patch in my previous comment makes things usable for me, but I don't know how to fix it properly (ie, without :silent!). Let me know if I'm doing something wrong or if you need more info.

.vimrc:

```
set nocompatible

call plug#begin('~/.vim/plugged')
Plug 'https://github.com/prabirshrestha/async.vim.git'
Plug 'https://github.com/prabirshrestha/vim-lsp.git'
call plug#end()

set completeopt=menuone,preview
"let g:lsp_diagnostics_float_delay = 120

function! s:on_lsp_buffer_enabled() abort
    setlocal omnifunc=lsp#complete
    setlocal signcolumn=yes
    nmap <buffer> gK <plug>(lsp-hover)
    nmap <buffer> gd <plug>(lsp-definition)
    nmap <buffer> gr <plug>(lsp-rename)
endfunction

augroup lsp_install
    au!
    autocmd User lsp_buffer_enabled call s:on_lsp_buffer_enabled()
augroup END

" python completion server
if executable('pyls')
    " pip3 install python-language-server
    au User lsp_setup call lsp#register_server({
                \ 'name': 'pyls',
                \ 'priority': 5,
                \ 'cmd': {server_info->['pyls']},
                \ 'whitelist': ['python'],
                \ })
endif
```

**Comment by randomgeek78** [comment]
I tested this on both `nvim (v0.5.0-566-g5f0a1b04c)` and `VIM - Vi IMproved 8.2`. 

As mentioned in the comments, `pclose` is used to close the popup window on nvim. 

`pclose` seems only effective when the Preview window is not open. Using the minimal configuration above, I can trigger the Preview window using <c-x><c-o> say after a `print|`. 

Say I don't trigger the popup using <c-x><c-o> and only trigger the popup/float by pressing `a` in `Normal` mode when curzor is on `(` in `print(`. In this case, `pclose` is effective. 

`nvim_win_close(s:winid, 0)` seems to properly fix the problem in both cases. I suggest we consider replacing `pclose` with the `nvim_win_close(s:winid, 0)`. The minimum diff against git commit `327b9d12` is as follows (I made a few modifications than what was suggested by @cristoper -- hope he does not mind):

```
index 0c09a45..38c97f5 100644
--- a/autoload/lsp/ui/vim/output.vim
+++ b/autoload/lsp/ui/vim/output.vim
@@ -19,6 +19,8 @@ function! lsp#ui#vim#output#closepreview() abort
     "works)
     if s:use_vim_popup && s:winid
         call popup_close(s:winid)
+    elseif s:use_nvim_float && s:wind
+        call nvim_win_close(s:winid, 0)
     else
         pclose
     endif
```

Given the edit is so small, I thought I would just past it here. Let me know if you prefer me making a PR.
